demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,626 @@
+    requestHeaders.pBuffer = userBuffer;
+    requestHeaders.bufferLen = USER_BUFFER_LENGTH;
+
+    /* Initialize the response object. The same buffer used for storing
+     * request headers is reused here. */
+    response.pBuffer = userBuffer;
+    response.bufferLen = USER_BUFFER_LENGTH;
+
+    LogInfo( ( "Getting file object size from host..." ) );
+
+    httpStatus = HTTPClient_InitializeRequestHeaders( &requestHeaders,
+                                                      &requestInfo );
+
+    if( httpStatus == HTTP_SUCCESS )
+    {
+        /* Add the header to get bytes=0-0. S3 will respond with a Content-Range
+         * header that contains the size of the file in it. This header will look
+         * like: "Content-Range: bytes 0-0/FILESIZE". The body will have a single
+         * byte that we are ignoring. */
+        httpStatus = HTTPClient_AddRangeHeader( &requestHeaders, 0, 0 );
+    }
+
+    if( httpStatus == HTTP_SUCCESS )
+    {
+        /* Send the request and receive the response. */
+        httpStatus = HTTPClient_Send( pTransportInterface,
+                                      &requestHeaders,
+                                      NULL,
+                                      0,
+                                      &response,
+                                      0 );
+    }
+
+    if( httpStatus == HTTP_SUCCESS )
+    {
+        LogInfo( ( "Received HTTP response from %s%s...",
+                   pHost, pPath ) );
+        LogInfo( ( "Response Headers:\n%.*s",
+                   ( int32_t ) response.headersLen,
+                   response.pHeaders ) );
+        LogInfo( ( "Response Status:\n%u",
+                   response.statusCode ) );
+        LogInfo( ( "Response Body:\n%.*s\n",
+                   ( int32_t ) response.bodyLen,
+                   response.pBody ) );
+
+        httpStatus = HTTPClient_ReadHeader( &response,
+                                            ( char * ) HTTP_CONTENT_RANGE_HEADER_FIELD,
+                                            ( size_t ) HTTP_CONTENT_RANGE_HEADER_FIELD_LENGTH,
+                                            ( const char ** ) &contentRangeValStr,
+                                            &contentRangeValStrLength );
+    }
+    else
+    {
+        LogError( ( "Failed to send HTTP %s request to %s%s: Error=%s.",
+                    pMethod, pHost, pPath, HTTPClient_strerror( httpStatus ) ) );
+    }
+
+    if( httpStatus == HTTP_SUCCESS )
+    {
+        /* Parse the Content-Range header value to get the file size. */
+        pFileSizeStr = strstr( contentRangeValStr, "/" );
+
+        if( pFileSizeStr == NULL )
+        {
+            LogError( ( "'/' not present in Content-Range header value: %s.",
+                        contentRangeValStr ) );
+        }
+
+        pFileSizeStr += sizeof( char );
+        *pFileSize = ( size_t ) strtoul( pFileSizeStr, NULL, 10 );
+
+        if( ( *pFileSize == 0 ) || ( *pFileSize == UINT32_MAX ) )
```

Comment:
What is wrong if the file size is equal to `UINT32_MAX`? Should we instead have a mechanism for detecting overflow?

Reply:
The standard states that it will return the maximum long integer if it detects overflow.